File: src/main/java/com/derbysoft/tinyweb/TinyWeb.java
package com.derbysoft.tinyweb;

import java.util.List;
import java.util.Map;

/**
 * Created by xinliwang on 14-9-28.
 */
public class TinyWeb {
    private final Map<String, Controller> controllers;
    private final List<Filter> filters;

    public TinyWeb(Map<String,Controller> controllers,List<Filter> filters){
        this.controllers=controllers;
        this.filters=filters;
    }

    public HttpResponse handleRequest(HttpRequest httpRequest) {
        HttpRequest currentRequest=httpRequest;
        for (Filter filter : filters) {
            currentRequest=filter.doFilter(httpRequest);
        }
        Controller controller = controllers.get(currentRequest.getPath());
        if(controller!=null){
            return controller.handleRequest(currentRequest);
        }
        return null;
    }
}

File: src/main/scala/com/derbysoft/fn/tinweb/TinyWeb.scala
package com.derbysoft.fn.tinweb

import com.derbysoft.fn.tinweb.base.{HttpRequest, HttpResponse}
import com.derbysoft.fn.tinweb.controller.Controller

/**
 * Created by xinliwang on 14-9-29.
 */
class TinyWeb(filters: List[(HttpRequest) => HttpRequest],
              controllers: Map[String, Controller]) {
  def handleRequest(request: HttpRequest): Option[HttpResponse] = {
   val composedFilter= filters.reverse.reduceLeft((composed, next) =>
     composed compose next)
    val filterRequest:HttpRequest=composedFilter(request)
    val controller = controllers.get(filterRequest.path);
    controller map (controller => controller.handleRequest(filterRequest))

  }


}
